clippy_lints/src/rebind_fn_arg_as_mut.rs

```diff
@@ -0,0 +1,92 @@
+use crate::utils::{snippet, span_lint_and_then};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{
+    BindingAnnotation, ExprKind, ImplItem, ImplItemKind, Item, ItemKind, Local, Node, PatKind, QPath, TraitFn,
+    TraitItem, TraitItemKind,
+};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for function arguments declared as not mutable and
+    /// later rebound as mutable.
+    ///
+    /// **Why is this bad?** It can be easily improved by just declaring the function
+    /// argument as mutable and removing the unnecessary assignment.
+    ///
+    /// **Known problems:** The function argument might have been shadowed by another
+    /// value before the assignment.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// fn foo_bad(bar: Vec<i32>) -> Vec<i32> {
+    ///     let mut bar = bar;
+    ///     bar.push(42);
+    ///     bar
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn foo(mut bar: Vec<i32>) -> Vec<i32> {
+    ///     bar.push(42);
+    ///     bar
+    /// }
+    /// ```
+    pub REBIND_FN_ARG_AS_MUT,
+    style,
```

Comment:
@ebroto what do you think of the lint level for this?

Reply:
I think this should be a `style` lint, but IMO we should:
* Check that the scope of the linted binding is the same as the scope of the parameter (so, the whole function body)
* Fix the known problem

If we want to leave these cases for a follow-up, I would add the lint to `nursery` for the time being.

Reply:
I think I can address your first suggestion this weekend.

As to the second, if by fixing you mean generating an applicable suggestion, I only didn't do it because the output looked awful: a suggestion to remove an entire line just prints out a blank that might confuse the user if they don't pay attention to the line number. If it's possible to make it better than that, let me know!

Reply:
Sorry for not being precise. I meant to avoid the false positive when the parameter has been shadowed.

Reply:
Oh, right. But actually, I think that is fixed by my new impl. I'm comparing the resolution id for the variable (path) with the HIR id of the parameter's pattern. The thing is, I'm starting to think we should expand this lint to cover not only function parameters but any binding. Is there some case in which `let mut x = x;` is actually desired?

Reply:
I believe that is already covered by the shadow_same lint.